protocols/relay/src/behaviour/handler.rs

```diff
@@ -396,149 +402,37 @@ impl Handler {
-                        Event::ReservationReqReceived {
-                            inbound_reservation_req,
-                            endpoint: self.endpoint.clone(),
-                            renewed: self.active_reservation.is_some(),
-                        },
-                    ));
-            }
-            inbound_hop::Req::Connect(inbound_circuit_req) => {
-                self.queued_events
-                    .push_back(ConnectionHandlerEvent::NotifyBehaviour(
-                        Event::CircuitReqReceived {
-                            inbound_circuit_req,
-                            endpoint: self.endpoint.clone(),
-                        },
-                    ));
-            }
-        }
+    fn on_fully_negotiated_inbound(&self, stream: Stream) {
+        self.protocol_futs.push(
+            inbound_hop::process_inbound_request(
+                stream,
+                self.config.reservation_duration,
+                self.config.max_circuit_duration,
+                self.config.max_circuit_bytes,
+                self.endpoint.clone(),
+                self.active_reservation.is_some(),
+            )
+            .boxed(),
+        );
     }
 
-    fn on_fully_negotiated_outbound(
-        &mut self,
-        FullyNegotiatedOutbound {
-            protocol: (dst_stream, dst_pending_data),
-            info: outbound_open_info,
-        }: FullyNegotiatedOutbound<
-            <Self as ConnectionHandler>::OutboundProtocol,
-            <Self as ConnectionHandler>::OutboundOpenInfo,
-        >,
-    ) {
-        let OutboundOpenInfo {
-            circuit_id,
-            inbound_circuit_req,
-            src_peer_id,
-            src_connection_id,
-        } = outbound_open_info;
+    fn on_fully_negotiated_outbound(&mut self, stream: Stream) {
+        let stop_command = self
+            .stop_requested_streams
+            .pop_front()
+            .expect("opened a stream without a pending stop command");
+
         let (tx, rx) = oneshot::channel();
         self.alive_lend_out_substreams.push(rx);
 
-        self.queued_events
-            .push_back(ConnectionHandlerEvent::NotifyBehaviour(
-                Event::OutboundConnectNegotiated {
-                    circuit_id,
-                    src_peer_id,
-                    src_connection_id,
-                    inbound_circuit_req,
-                    dst_handler_notifier: tx,
-                    dst_stream,
-                    dst_pending_data,
-                },
-            ));
-    }
-
-    fn on_listen_upgrade_error(
-        &mut self,
-        ListenUpgradeError {
-            error: inbound_hop::UpgradeError::Fatal(error),
-            ..
-        }: ListenUpgradeError<
-            <Self as ConnectionHandler>::InboundOpenInfo,
-            <Self as ConnectionHandler>::InboundProtocol,
-        >,
-    ) {
-        self.pending_error = Some(StreamUpgradeError::Apply(Either::Left(error)));
-    }
-
-    fn on_dial_upgrade_error(
```

Comment:
I think this function should largely stay as is! We just won't have an `UpgradeError` any more but we should still handle all the other variants.

Reply:
Yes, I agree, but all errors happen inside the `outbound_stop::handle_stop_message_response` method, so I carefully moved errors handling to this method.

Reply:
> but all errors happen inside the `outbound_stop::handle_stop_message_response` method

I don't think that is quite true. `DialUpgradeError` can still return things like `StreamUpgradeError::Timeout` which is currently unhandled. See line 802.

Reply:
> > but all errors happen inside the `outbound_stop::handle_stop_message_response` method
> 
> I don't think that is quite true. `DialUpgradeError` can still return things like `StreamUpgradeError::Timeout` which is currently unhandled. See line 802.

Yes, You are right. Returned the `on_dial_upgrade_error` function.